### PR TITLE
ENH: linalg: add batch support for matrix -> scalar funcs

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Follow-up to gh-21462
Toward gh-21466

#### What does this implement/fix?
This adds support for n-dimensional batch input to `linalg.expm_cond`, `linalg.issymmetric`, and `linalg.ishermitian` - the remaining functions that accept a matrix and return a scalar.

It also creates a `test_batch` file where we can keep all the batch support tests.

### CODE REVIEW

File: scipy/linalg/tests/test_batch.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import numpy as np
+from numpy.testing import assert_allclose
+from scipy import linalg
+
+
+class TestMatrixInScalarOut:
+
+    def batch_test(self, fun, args=(), kwargs=None, dtype=np.float64,
```
Comment (mdhaber) [inline scipy/linalg/tests/test_batch.py:8]: This structure will let us parametrized over `args`/`kwargs`/`dtype` for each function separately. We can add more generality as needed.

File: scipy/linalg/tests/test_batch.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+        rng = np.random.default_rng(seed)
+        A = rng.random(batch_shape + core_shape).astype(dtype)
+        res = fun(A, *args, **kwargs)
+
+        for i in range(batch_shape[0]):
+            for j in range(batch_shape[1]):
+                ref = fun(A[i, j], *args, **kwargs)
+                assert_allclose(res[i, j], ref)
+
+    def test_expm_cond(self):
```
Comment (mdhaber) [inline scipy/linalg/tests/test_batch.py:20]: Could parametrized over `dtype` if desired. 
We already tested in `eig` that the decorator passes other arguments in as expected, so I don't really feel a need to test with `check_finite` here. I tested locally and it works as expected.

**Comment by j-bowhay** [comment]
> * `det` and `norm` are already done
> * `issymmetric` and `ishermitian` are written in Cython
> 
> How should I treat functions that have no Python code? Import them in the `__init__` and wrap them there?

Since `issymmetric` and `ishermitian` are standard `def` functions is it not possible to apply the decorator directly in the `.pyx` file?

**Comment by mdhaber** [comment]
You know I didn't even look. I'll do that here, thanks!

File: scipy/linalg/tests/test_batch.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        # test_expm_cond doesn't need symmetric/hermitian matrices, and
+        # test_issymmetric doesn't need hermitian matrices, but it doesn't hurt.
+        A = get_nearly_hermitian(batch_shape + core_shape, dtype, 3e-4, rng)
+
+        res = fun(A, *args, **kwargs)
+
+        for i in range(batch_shape[0]):
+            for j in range(batch_shape[1]):
+                ref = fun(A[i, j], *args, **kwargs)
+                assert_allclose(res[i, j], ref, strict=True)
```
Comment (mdhaber) [inline scipy/linalg/tests/test_batch.py:37]: : (
```suggestion
                assert_allclose(res[i, j], ref)
```

Comment (mdhaber) [reply scipy/linalg/tests/test_batch.py:37]: Just another reason to get to NumPy 2.0

**Comment by mdhaber** [comment]
Thanks @j-bowhay. Now this PR takes care of all three functions.

**Review by j-bowhay — APPROVED** [review]
Thanks! It very nice how minimal the change is using the decorator

**Pull request merged**